### ISSUE TITLE
Client hangs on connect

### ISSUE THREAD

**Issue by TroelsL** [chkr1011/MQTTnet#279]
I'm trying to port parts of a javascript project to .NET Core, but when running it, it immediately disconnects, then hangs the thread. It is entirely likely that I am doing something wrong, but I can't seem pinpoint the issue. I never reach any of the event handlers.

        private async Task ConnectToQueue()
        {
            var factory = new MqttFactory();

            var mqClient = factory.CreateMqttClient(new Logger());
            var options = new MqttClientOptionsBuilder()
                .WithClientId(_authInfo.MqttClientId)
                .WithTcpServer(_authInfo.MqttEndpoint)
                .WithTls(certificates: new[] { _rawCertificate })
                .WithCommunicationTimeout(TimeSpan.FromHours(24))
                .Build();

            mqClient.ApplicationMessageReceived += Client_ApplicationMessageReceived;
            mqClient.Connected += MqClient_Connected;
            var connection = await mqClient.ConnectAsync(options);

        }

        private void MqClient_Connected(object sender, MqttClientConnectedEventArgs e)
        {
            var client = sender as IMqttClient;
            client.SubscribeAsync($"DB510/{_productInfo.MacAddress}/commandOut");
            client.SubscribeAsync($"DB510/{_productInfo.MacAddress}/commandIn");
        }

        private void Client_ApplicationMessageReceived(object sender, MqttApplicationMessageReceivedEventArgs e)
        {
            Console.WriteLine(Encoding.UTF8.GetString(e.ApplicationMessage.Payload));
        }

Above yields the following log messages and hangs at ConnectAsync.

    VER: Trying to connect with server.
    VER: Connecting [Timeout=1.00:00:00]
    VER: Connection with server established.
    VER: Start receiving packets.
    VER: TX >>> Connect: [ClientId=<ClientId>] [Username=] [Password=] [KeepAlivePeriod=15] [CleanSession=True] [Timeout=1.00:00:00]
    VER: Disconnecting [Timeout=1.00:00:00]
    VER: Disconnected from adapter.
    INF: Disconnected.
    VER: Stopped receiving packets.

**Comment by TroelsL** [comment]
I'll give it a try later today, but I don't think it'll make a difference as the thread hangs on the await ConnectAsync line, thus never reaching the following lines where the while loop is.

**Comment by TroelsL** [comment]
Indeed, with the sample code I am stuck in an infinite loop as it fails to connect. How can I find out what the cause of the immediate disconnect is?

`VER: Trying to connect with server.
VER: Connecting [Timeout=1.00:00:00]
VER: Connection with server established.
VER: Start receiving packets.
VER: TX >>> Connect: [ClientId=<CID>] [Username=] [Password=] [KeepAlivePeriod=15] [CleanSession=True] [Timeout=1.00:00:00]
VER: Disconnecting [Timeout=1.00:00:00]
VER: Disconnected from adapter.
INF: Disconnected.
### DISCONNECTED FROM SERVER ###
VER: Stopped receiving packets.
VER: Trying to connect with server.
VER: Connecting [Timeout=1.00:00:00]
VER: Connection with server established.
VER: Start receiving packets.
VER: TX >>> Connect: [ClientId=<CID>] [Username=] [Password=] [KeepAlivePeriod=15] [CleanSession=True] [Timeout=1.00:00:00]
VER: Disconnecting [Timeout=1.00:00:00]
VER: Disconnected from adapter.
INF: Disconnected.
### DISCONNECTED FROM SERVER ###
VER: Stopped receiving packets.
VER: Trying to connect with server.
VER: Connecting [Timeout=1.00:00:00]
VER: Connection with server established.
VER: Start receiving packets.
VER: TX >>> Connect: [ClientId=<CID>] [Username=] [Password=] [KeepAlivePeriod=15] [CleanSession=True] [Timeout=1.00:00:00]
VER: Disconnecting [Timeout=1.00:00:00]
VER: Disconnected from adapter.
INF: Disconnected.
### DISCONNECTED FROM SERVER ###
VER: Stopped receiving packets.
VER: Trying to connect with server.
VER: Connecting [Timeout=1.00:00:00]
VER: Connection with server established.
VER: Start receiving packets.
VER: TX >>> Connect: [ClientId=<CID>] [Username=] [Password=] [KeepAlivePeriod=15] [CleanSession=True] [Timeout=1.00:00:00]
VER: Disconnecting [Timeout=1.00:00:00]
VER: Disconnected from adapter.
INF: Disconnected.
### DISCONNECTED FROM SERVER ###
VER: Stopped receiving packets.`

**Comment by TroelsL** [comment]
I have downloaded the source, and it hangs on the 24h timeout while trying to authenticate. I suspect a client certificate issue as there is no other auth mechanism other than client id. Is this possible to somehow get more information on from the client? I know the server works as it is possible to connect from node.js.

**Comment by chkr1011** [comment]
Which broker do you use? I also got such issues when using TLS with an unencrypted broker or disabled TLS with an encrypted broker.

**Comment by TroelsL** [comment]
@chkr1011 I'm not sure what's on the other end. I'm trying to integrate with my robot lawnmower. All I have to go by is some working javascript code from here:

https://github.com/MeisterTR/ioBroker.landroid-s/blob/master/lib/landroid-cloud-2.js

Right now, I'm trying to get a dump from the android app and compare it to what my C# implementation does.

**Comment by TroelsL** [comment]
@biapar: I do now own or operate the server. I am accessing a 3rd party system.

**Comment by TroelsL** [comment]
I finally managed to track the issue down, and it was a problem with the client certificate. While actually valid, it was not accepted by the server.

I don't know how the protocol responds to this, but if it were possible to add an error somehow, that would've saved me a lot of time :)

**Comment by TroelsL** [comment]
Based on what you said, I did some further investigations, and I've come to conclusion that I don't think the client is behaving in any way it shouldn't. It is the result of requesting a client certificate with a wrong authentication token and thus getting a valid certificate, but not for anything meaningful.
Basically, I'm subscribing to a queue that in my book should not exist, and instead of returning the equivalent of a 404 or some other error, it simply subscribes me to an unused queue.

Sorry for the confusion, this one was really difficult to track down. I had to install both the MQTTnet source and the source javascript library with all its dependencies.
